### DISCUSSION TITLE
"floating invalid" when using a generic metforcing.

### DISCUSSION THREAD

**Discussion by bhanu-magotra** [NASA-LIS/LISF#1295] [Q&A] [answered]
Hi, 
I am trying to run NoahMP3.6 with Hymap2  at 0.1 degrees and 1 hour time-step over the Indian subcontinent using two forcings:
1. A generic local precipitation at 0.25 degrees and daily time-step made to resemble LDT Generated forcing with following attributes:
```
        double lon(north_south, east_west) ;
                lon:_FillValue = NaN ;
        double lat(north_south, east_west) ;
                lat:_FillValue = -9999. ;
                lat:standard_name = "latitude" ;
                lat:units = "Degrees North" ;
        int64 time(time) ;
                time:time_increment = "86400" ;
                time:units = "days since 1980-11-02 00:00:00" ;
                time:calendar = "proleptic_gregorian" ;
        float Rainf_tavg(time, north_south, east_west) ;
                Rainf_tavg:_FillValue = -9999.f ;
                Rainf_tavg:missing_value = NaNf ;
                Rainf_tavg:standard_name = "Rainf" ;
                Rainf_tavg:units = "kg m-2 s-1" ;
                Rainf_tavg:coordinates = "lat lon" ;
```
2. MERRA2 as a global forcing that provides other variables. Also, the local precipitation is available on a smaller area compared to running LIS domain and hence MERRA2 is expected to provide the missing precipitation.

The lis config is configured to use MERRA2 and ldt-generated forcing as per following :
```
Number of met forcing sources:      2
Met forcing sources:                 "MERRA2"  "LDT-generated" 
Blending method for forcings:       "overlay"
Topographic correction method (met forcing):  "none" "none"
Enable spatial downscaling of precipitation:   0   0
Spatial interpolation method (met forcing):   "budget-bilinear"  "bilinear"
Temporal interpolation method (met forcing):  "none" "none"
Spatial upscaling method (met forcing):       "none"  "none"
```
While running the LIS in coldstart mode from 1981-01-01 onwards over complete domain, the run gets initiated and LIS successfully identifies LDT-generated forcing for 1 variable (Rainf_tavg) .
```
[INFO] Using MERRA2 forcing
 [INFO] MERRA2 forcing directory: ../../datasets/MERRA2
 [INFO] 
 [INFO] Using LDT-generated met forcing data
 [INFO] Generated metforcing directory: 
 /home/civil/phd/cez218606/LISF1/Ref_IMD/IMD_new
 [INFO] -- Obtain Forcing Dataset Parameters -- 
 [INFO] -- Parameters from forcing file: 
 /home/civil/phd/cez218606/LISF1/Ref_IMD/IMD_new/FORCING/198101/LDT_HIST_1981010
 10000.d01.nc
 [INFO] LDT-Generated Forcing Timestep:    86400.00    
 [INFO] LDT-Generated Forcing Number of Cols, Rows:          135         129
 [INFO] LDT-Generated Forcing Projection:  Latlon 
 [INFO] Allocating and Initializing Forcing Variables -- 
  ** LDT-generated variable being read in: Rainf_tavg
 [INFO] Number of LDT-generated forcing fields:            1
 [INFO] 
```
But stops abruptly at following : 
```
 [INFO] Reading MERRA-2 file (bookend,           2  ... 
 ./MERRA2/MERRA2_100/Y1981/M01/MERRA2_100.tavg1_2d_rad_Nx.19810102.nc4
  [ERR] Currently Forcing 1 File is NOT opened in first timestep ...
 [INFO] Reading in forcing file: 
 /home/civil/phd/cez218606/LISF1/Ref_IMD/IMD_new/FORCING/198101/LDT_HIST_1981010
 20000.d01.nc
  - Reading and spatially reprojecting forcing fields - 
  Done reading and spatially reprojecting forcing fields - 
```
The PBS error logs shows following message:

```
forrtl: error (65): floating invalid
Image              PC                Routine            Line        Source             
LIS                0000000001425A54  Unknown               Unknown  Unknown
libpthread-2.17.s  00002B9DC62BC630  Unknown               Unknown  Unknown
LIS                0000000000E7905C  noahmp_routines_3        6800  module_sf_noahmplsm_36.F90
LIS                0000000000E640B1  noahmp_routines_3        1005  module_sf_noahmplsm_36.F90
LIS                00000000010CC752  noahmp_driver_36_         558  noahmp_driver_36.F90
LIS                0000000000783473  noahmp36_main_            615  NoahMP36_main.F90
LIS                000000000065F920  lis_lsmmod_mp_lis         619  LIS_lsmMod.F90
LIS                00000000007156BC  lis_surfacemodelm         248  LIS_surfaceModelMod.F90
LIS                00000000012974E1  retrospective_run         198  retrospective_runMod.F90
LIS                0000000000D36FD9  MAIN__                     70  lisdrv.F90
LIS                0000000000416492  Unknown               Unknown  Unknown
libc-2.17.so       00002B9DC67ED555  __libc_start_main     Unknown  Unknown
LIS                00000000004163A9  Unknown               Unknown  Unknown
real 6.02
user 21.91
sys 1.68
```
The [line 6800](https://github.com/NASA-LIS/LISF/blob/c9ab9288e7de557defbe92a94b6ac7e52344405d/lis/surfacemodels/land/noahmp.3.6/module_sf_noahmplsm_36.F90#L6800) belongs to CANWATER module and [Line1005](https://github.com/NASA-LIS/LISF/blob/c9ab9288e7de557defbe92a94b6ac7e52344405d/lis/surfacemodels/land/noahmp.3.6/module_sf_noahmplsm_36.F90#L1005) belongs to Water Budget module
I have checked the local precipiation for any negative or infinity values but everything seems in order. Also, the runs are successfull if I used a much smaller land-locked domain but fails on the complete Indian Subcontinent grid which also includes coastal boundaries. Ideally, LIS should be able to tackle the missing or invalid local precip by substituting MERAA2 precip but that isn't happening. 

I am attaching relevant lisconfig and lislog outputs. Please note that all lislog.XXXX end at same output. I would appreciate any help regarding this @dmocko  , @karsenau .

Thanks. 
[fltInv.zip](https://github.com/NASA-LIS/LISF/files/10999601/fltInv.zip)

**Comment by bhanu-magotra** [comment]
After downscaling the local forcing to match lis input file grid resolution (0.1 degrees), we were able to get pass the initial error but the LIS surfacemodel outputs suggest that LIS is ingesting MERRA2 precipitation even in the overlay method.
Here are the lisconfig parameters:
```
Number of met forcing sources:      2
Met forcing sources:                 "MERRA2" "LDT-generated"
Blending method for forcings:       "overlay"
Topographic correction method (met forcing):  "none" "none"
Enable spatial downscaling of precipitation:   0   0
Spatial interpolation method (met forcing):   "budget-bilinear"  "budget-bilinear"
Temporal interpolation method (met forcing):  "linear" "linear"
Spatial upscaling method (met forcing):       "none"  "none"
```
The lislog.XXXX shows LDT-generated Rainf_tavg being read but output precipitation matches to standalone MERRA2 outputs, suggesting that LDT-generated precipitation is not being used.
Here we can see LDT generated forcing being read in : 
```
 [INFO] Using LDT-generated met forcing data
 [INFO] Generated metforcing directory: 
 /home/civil/phd/cez198621/msaharia/ved/LISF1/Ref_IMD/IMD_new
 [INFO] -- Obtain Forcing Dataset Parameters -- 
 [INFO] -- Parameters from forcing file: 
 /home/civil/phd/cez198621/msaharia/ved/LISF1/Ref_IMD/IMD_new/FORCING/198101/LDT
 _HIST_198101010000.d01.nc
 [INFO] LDT-Generated Forcing Timestep:    86400.00    
 [INFO] LDT-Generated Forcing Number of Cols, Rows:          297         290
 [INFO] LDT-Generated Forcing Projection:  Latlon 
 [INFO] Allocating and Initializing Forcing Variables -- 
  ** LDT-generated variable being read in: Rainf_tavg
 [INFO] Number of LDT-generated forcing fields:            1
```
And here are the two outputs compared for the TotalPrecip_tavg values on 1981-01-05:
```
LDT-GEN + MERRA2 Overlay = 0.0  to 0.000553
MERRA2 Standalone = 0.0 to 0.000553
```
Any ideas about how can we diagnose the issue?
Regards.

**Reply by karsenau** [reply]
@bhanu-magotra:

You could run LIS with just the LDT-generated met forcing data that you created and make sure that the data are being read in and written out properly, putting in just 1 metforcing source and run without the LSM to test the read/write of the files you created.

**Reply by bhanu-magotra** [reply]
Hi @karsenau ,
Thanks for the suggestion.
I tried running LIS with template lsm and accordingly turned all fields off both in inputs and outputs except precipitation. The output TotalPrecip_tavg are not as expected. They are all NaN values.
Here is my config:
```
#Overall driver options:
Running mode:                       "retrospective"
Map projection of the LIS domain:   "latlon"
Number of nests:                    1
Number of surface model types:      1
Surface model types:                "LSM"
Surface model output interval:      "1da"         # "1da", "1hr", "1mo
Land surface model:                 "none"
Number of ensembles per tile:       1
Number of met forcing sources:      1
Met forcing sources:                "LDT-generated" 
Blending method for forcings:       "overlay"
Topographic correction method (met forcing):  "none" "none"
Enable spatial downscaling of precipitation:   0   0
Spatial interpolation method (met forcing):   "budget-bilinear"  "budget-bilinear"
Temporal interpolation method (met forcing):  "linear" "linear"
Spatial upscaling method (met forcing):       "none"  "none"

```
Am I doing this correctly? If yes, how do I further diagnose the issue. 
Please note that if I plot the LDT-generated source precipitation, it appears to be the usual and as expected.

Regards.

**Reply by dmocko** [reply]
Hi @bhanu-magotra 

Yes, you appear to be doing this correctly to diagnose the problem.

Can you print/write the data right after it is read in?  It will help diagnose if it is being read correctly.

https://github.com/NASA-LIS/LISF/blob/master/lis/metforcing/genMetForc/metForcGen_VariablesMod.F90#L458-L472

Around Line 483 in the above code, add these lines:

print *,input_var
stop

Re-compile and re-run.  Does it show the correct precip values?  If not, we'll have to go further back to find the problem.  If it does show correct values, then we can trace it forward to see what's going on.

**Reply by bhanu-magotra** [reply]
Hi @dmocko , 
Thank you for the suggestion.
I tried it and the print output is showing valid values in between many NaNs. I am attaching the output for your consideration. 
Please advise further.
Thanks.
[printVar.txt](https://github.com/NASA-LIS/LISF/files/11032326/printVar.txt)
[lislog.txt](https://github.com/NASA-LIS/LISF/files/11032344/lislog.txt)

**Comment by dmocko** [comment]
Can you please attach the result of "ncdump" on this file:

/home/civil/phd/cez198621/msaharia/ved/LISF1/Ref_IMD/IMD_new/FORCING/198101/LDT_HIST_198101020000.d01.nc

**Reply by bhanu-magotra** [reply]
Hi @dmocko ,
I am attaching the 'ncdump' of concerned file.
[ncdump.txt](https://github.com/NASA-LIS/LISF/files/11036377/ncdump.txt)

**Comment by dmocko** [comment, marked as answer]
Hi - it looks like the data is being read in correctly by LIS.  I see plenty of NaNs in the precip field, along with reasonable values.

Please look in the LDT code for generating this forcing to see why NaNs are being written into this file (instead of undefined values).

You can use print statements, etc.  We have limited time to help, and given your level of expertise, you should be able to diagnose why this is happening.  Please examine the code and use print statements and/or a debugger to follow why these NaNs are being written.

**Reply by bhanu-magotra** [reply]
Hi, 
Since we use a python code to make local forcing resemble a LDT-generated one, we fill NaNs as float(np.nan).  Isn't LIS expecting NaNf for overlay tasks?
What should be the ideal way to define NaNs for generic met-forcing when preparing through a python code, so that LIS can do the overlay?
Thanks.

**Reply by dmocko** [reply]
Ah - thanks for the explanation.  Glad it's not the LDT code that is incorrectly generating these NaNs.  It's your python code.

Instead of defining missing value as NaNs, how about setting them to a specific value (e.g., -9999.0)?  It should be the same undefined value that is set in your lis.config.

**Reply by bhanu-magotra** [reply]
Alright, let me try a specific value.

Thanks.

**Reply by bhanu-magotra** [reply]
@dmocko , I tried with -9999.0 but the end result is still same.

**Reply by dmocko** [reply]
These values in your netCDF file are incorrect:

```
                :SOUTH_WEST_CORNER_LAT = 8.1 ;
                :SOUTH_WEST_CORNER_LON = 37.05 ;
                :NORTH_EAST_CORNER_LAT = 68.05 ;
                :NORTH_EAST_CORNER_LON = 97.65 ;
```

According to the "lon" and "lat" values of your netCDF file, these values should be:

```
                :SOUTH_WEST_CORNER_LAT = 8.1 ;
                :SOUTH_WEST_CORNER_LON = 68.05 ;
                :NORTH_EAST_CORNER_LAT = 37.0 ;
                :NORTH_EAST_CORNER_LON = 97.65 ;
```

Please re-generate the forcing files with these corrected global attributes.  LIS reads these lines to determine the spatial extent of the forcing.

Also, please attach your ldt.config file, which sets your run domain.

**Reply by bhanu-magotra** [reply]
The issue is resolved. The wrong global attribute was the final culprit.

Thank you @dmocko @karsenau

**Reply by karsenau** [reply]
Glad you were able to find the issue.
